### PR TITLE
feat(Synthwave): add Synthwave #33

### PR DESCRIPTION
**Pull request by andretchen0**
synthwave

Here's at attempt at an infinite procedural Synthwave landscape. 

## Bug?

The demo uses some post-processing. For that, the renderer is required. I tried to get the renderer like the post-processing demo, but the post-processing demo doesn't work. [I opened an issue.](https://github.com/Tresjs/playground/issues/68)

Instead, I hacked a solution together using `onAfterRender` on a dummy element that is subsequently removed. `onAfterRender` is a callback that gets the renderer as an argument.

If anyone knows how to do this without resorting to hacks, please let me know and I'll fix it.

## Feedback

Feedback welcome. At the moment, particularly:

* visuals
* how to better show off Tres/Cientos
* how to fix the bug above

### CODE REVIEW

**Comment by andretchen0** [comment]
> @andretchen0 is still in Draft or its ready for preview? By the way, I have a proposal for you 😉. Check your discord DM

Still in draft. I wanted to get your feedback about:

* visuals
* how to better show off Tres/Cientos
* how to fix the bug above

**Comment by alvarosabu** [comment]
Hi, @andretchen0 to answer your questions:

Postprocessing @Tinoooo is working on https://github.com/Tresjs/post-processing so you can use it but probably we are missing some of the effects you need. Take a look if you like, that would solve the bug but we would need fo add the FFXA effect

Should we add `onAfterRender` call back on the `<TresCanvas @after-render="onAfterRender" />`?

Wdyt?

**Comment by Tinoooo** [comment]
@andretchen0 You can get the renderer by using [useTresContext](https://github.com/Tresjs/tres/blob/20ffa4bbc8d8a45643c8d4368a91cf68513b94c2/src/composables/useTresContextProvider/index.ts#L92) like so:

```ts
const { renderer } = useTresContext()
```

**Comment by alvarosabu** [comment]
> @andretchen0 You can get the renderer by using [useTresContext](https://github.com/Tresjs/tres/blob/20ffa4bbc8d8a45643c8d4368a91cf68513b94c2/src/composables/useTresContextProvider/index.ts#L92) like so:

Yep but only within a subcomponent of TresCanvas

**Comment by andretchen0** [comment]
@alvarosabu @Tinoooo 

Thanks for the pointers.

**Comment by andretchen0** [comment]
> Hi, @andretchen0 to answer your questions:
> 
> Postprocessing @Tinoooo is working on https://github.com/Tresjs/post-processing so you can use it but probably we are missing some of the effects you need. Take a look if you like, that would solve the bug but we would need fo add the FFXA effect

Sounds good. I'm still playing around with post-processing.

> Should we add `onAfterRender` call back on the `<TresCanvas @after-render="onAfterRender" />`?

I don't need it here. I can't say if it's worth expanding the API.

Sorry. That's probably not very helpful! ;)

**Comment by Tinoooo** [comment]
> > @andretchen0 You can get the renderer by using [useTresContext](https://github.com/Tresjs/tres/blob/20ffa4bbc8d8a45643c8d4368a91cf68513b94c2/src/composables/useTresContextProvider/index.ts#L92) like so:
> 
> Yep but only within a subcomponent of TresCanvas

You are right. This will change with https://github.com/Tresjs/tres/issues/403.

**Comment by andretchen0** [comment]
Reviews welcome.

There's now a "day/night" effect. Just in case there are rendering inconsistencies, it should look like this:

https://github.com/Tresjs/playground/assets/20469369/f86d8b47-34b6-456b-9f02-8a35ace21818

**Comment by alvarosabu** [comment]
Hey @andretchen0 is the demo finished?

**Comment by andretchen0** [comment]
> Hey @andretchen0 is the demo finished?

Yes, unless you have some suggestions.

File: components/content/synthwave/components/postprocessing.vue [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<script lang="ts" setup>
+import { WebGLRenderer, Camera, Scene } from 'three'
+import { BloomEffect, ScanlineEffect, EffectComposer, RenderPass, EffectPass } from 'postprocessing'
+import { useWindowSize } from '@vueuse/core'
+let effectComposer: EffectComposer
+
+function setup(renderer: WebGLRenderer, scene: Scene, camera: Camera) {
```
Comment (alvarosabu) [inline components/content/synthwave/components/postprocessing.vue:7]: Is `@tresjs/post-processing` not covering the use-case? What would be missing?

**Review by alvarosabu — APPROVED** [review]
I think it looks amazing, clean use of shaders. The effect is amazing.

You know what could put the cherry on top? Finding a CC0 synthwave song and add a "Music On/Off" button, wdyt?

**Comment by andretchen0** [comment]
> I think it looks amazing, clean use of shaders. The effect is so dope.
> 
> You know what could put the cherry on top? Finding a CC0 synthwave song and add a "Music On/Off" button, wdyt?

[What do you think of this one?
](https://freesound.org/people/vlad.zaichyk/sounds/495329/) It's CC0. The Freesound player fades in at the beginning, but the loop is actually pretty clean.

Even though it's CC0, I'll add an attribution link if we use it, if that's ok by you. Artists don't get enough love!

File: components/content/synthwave/components/postprocessing.vue (hunk 1 above)
Comment (andretchen0) [reply components/content/synthwave/components/postprocessing.vue:7]: There's no reason except that I hadn't used `post-processing`. :) I'll switch things over.

**Comment by alvarosabu** [comment]
> Even though it's CC0, I'll add an attribution link if we use it, if that's ok by you. Artists don't get enough love!

Absolutely, you can add it with :the-info: markdown 

Like the song, lets go for it

**Comment by alvarosabu** [comment]
@andretchen0  I just merged the asset in the other repo

**Comment by andretchen0** [comment]
I think this is done and can be merged, unless you'd like changes here:

## Post-processing

Earlier, I had just reused code found in the ["Manual Post-Processing" demo.](https://deploy-preview-69--tresjs-playground.netlify.app/experiments/post-processing-manual) It doesn't use template components.

I tried to use the template components, but I ended up getting errors. Is there some working code I can look at with the template components?

## Lines

There was some weirdness with the geometry/post-processing combo in Chrome. See the "peaks":

https://github.com/Tresjs/playground/assets/20469369/4b6c471e-ab58-4af3-a3ac-567205118869

So I turned off the lines in the distance and synced them up to the beat of the music, which I think is a better effect anyway.

**Comment by alvarosabu** [comment]
Awesome, I'm gonna quickly fix the sound on mobile for some reason is muted and the button is not working.

Gonna share it today on socials, by the way @andretchen0 do you have twitter to add you to the post Credit?

**Comment by andretchen0** [comment]
> Awesome, I'm gonna quickly fix the sound on mobile for some reason is muted and the button is not working after merging it

Thanks!

> do you have Twitter to add you to the post Credit?

No. I'm terrible at self-promo.

I'll put it on the Discord though.

**Pull request merged**